rasa_core_sdk/forms.py

```diff
@@ -5,176 +5,307 @@
+from rasa_core_sdk.events import SlotSet, Form
 
 logger = logging.getLogger(__name__)
 
-# this slot is used to store information needed
-# to do the form handling, needs to be part
-# of the domain
-FORM_SLOT_NAME = "requested_slot"
-
-
-class FormField(object):
-
-    # noinspection PyMethodMayBeStatic
-    def validate(self, value):
-        """Check if extracted value for a requested slot is valid.
-
-        Users should override this to implement custom validation logic,
-        returning None indicates a negative validation result, and the slot
-        will not be set.
-        """
-        return value
-
-
-class EntityFormField(FormField):
-
-    def __init__(self, entity_name, slot_name):
-        self.entity_name = entity_name
-        self.slot_name = slot_name
-
-    def extract(self, tracker):
-        # type: (Tracker) -> List[EventType]
-
-        value = next(tracker.get_latest_entity_values(self.entity_name), None)
-        validated = self.validate(value)
-        if validated is not None:
-            return [SlotSet(self.slot_name, validated)]
-        else:
-            return []
-
+if typing.TYPE_CHECKING:
+    from rasa_core_sdk import Tracker
+    from rasa_core_sdk.executor import CollectingDispatcher
 
-class BooleanFormField(FormField):
-    """A form field that prompts the user for a yes or no answer.
-    The interpreter should map positive and negative answers to
-    the intents ``affirm_intent`` and ``deny_intent``.
-    """
-
-    def __init__(self, slot_name, affirm_intent, deny_intent):
-        self.slot_name = slot_name
-        self.affirm_intent = affirm_intent
-        self.deny_intent = deny_intent
-
-    def extract(self, tracker):
-        # type: (Tracker) -> List[EventType]
-
-        intent = tracker.latest_message.get("intent", {}).get("name")
-        if intent == self.affirm_intent:
-            value = True
-        elif intent == self.deny_intent:
-            value = False
-        else:
-            return []
-
-        return [SlotSet(self.slot_name, value)]
-
-
-class FreeTextFormField(FormField):
-
-    def __init__(self, slot_name):
-        self.slot_name = slot_name
-
-    def extract(self, tracker):
-        # type: (Tracker) -> List[EventType]
-
-        validated = self.validate(tracker.latest_message.get("text"))
-        if validated is not None:
-            return [SlotSet(self.slot_name, validated)]
-        return []
+# this slot is used to store information needed
+# to do the form handling
+REQUESTED_SLOT = "requested_slot"
 
 
 class FormAction(Action):
+
     def name(self):
         # type: () -> Text
-        """Unique identifier of this form action."""
-
-        raise NotImplementedError
+        """Unique identifier of the form"""
 
-    RANDOMIZE = True
+        raise NotImplementedError("A form must implement a name")
 
     @staticmethod
-    def required_fields():
-        return []
+    def required_slots(tracker):
```

Comment:
can we add a comment to this that you can modify this function to request conditional slots as well? so like not just required ones, but if certain slots are filled you can ignore some other slots

Reply:
added a phrase